### ISSUE TITLE
Design and runtime question: make grain calls from a DLL

### ISSUE THREAD

**Issue by fantasticfears** [dotnet/orleans#6059]
I'm building some flexible commands that can make use of my grain interfaces where these commands shipped by a DLL other than silo. The main goal is to have a lot of those commands without even bothering updating slioes.

So in my silo, I inject a proxy to build such commands by name. The proxy looks for a dll in a repository for the symbol that given on runtime.
`services.AddScoped<IActionProxy, ActionProxy>();`
And when building these commands, a `GrainFactory` is injected. Its constructor has the following signature: `public FooAction(IConfiguration config, IGrainFactory grainFactory, ILogger logger)`.
But it doesn't send my invocation to the actual grain. Moreover, I don't realize if they reference the correct address. I assume those actions (commands) didn't reference grains on the silo.

I haven't read much Orleans code but keep wondering if I am making the right choices when designing these actions. So for reference, I will detail a high-level design about how it should work.
Basically a grain will do something and asks the proxy for a certain action. This should work fine since a grain can get the service. Then action was built with injected GrainFactory. This action will try to call the grain from GrainFactory.

I am skeptical about the very last step in my design. I would appreciate some pointers.

Orleans version: 3.0.0-rc2
.NET Core 3 on Windows

**Comment by jason-bragg** [comment]
"The main goal is to have a lot of those commands without even bothering updating silos."

From your pattern it's not clear if runtime loading of grain assemblies is expected.  If so, consider testing this aspect early, as I don't think Orleans handles that.  Currently, all grain assemblies are loaded at silo start and I know of no dynamic loading support.

Also, possibly relevant to flexible command patterns over Orleans, you may want to consider using grain extensions for cross cutting concerns.  They are not yet documented, but please see #5905 for some details of this.

**Comment by fantasticfears** [comment]
> From your pattern it's not clear if runtime loading of grain assemblies is expected. If so, consider testing this aspect early, as I don't think Orleans handles that. Currently, all grain assemblies are loaded at silo start and I know of no dynamic loading support.

Loaded code from DLL is running  For the proxy, the passed `GrainFactory` contains an InsideRuntimeClient so it knows where to find the silo. I made some progress in debugging as everything works in the debugger but not when it runs without breakpoints. Investigating

**Comment by fantasticfears** [comment]
Okay, the action has the right runtime. It's that there are multiple threads running different activations that result in my issues. Thanks for the help.
